### DISCUSSION TITLE
Multiple Cache-Control header

### DISCUSSION THREAD

**Discussion by magicW** [symfony/symfony#54369] [Q&A] [answered]
I'm not sure if I'm doing this right. I'm working on our cache-control header and I saw that we send two different Headers.
The first is set because of the Session-usage in AbstractSessionHandler:

```php
public function open($savePath, $sessionName)
    {
        $this->sessionName = $sessionName;
        if (!headers_sent() && !\ini_get('session.cache_limiter') && '0' !== \ini_get('session.cache_limiter')) {
            header(sprintf('Cache-Control: max-age=%d, private, must-revalidate', 60 * (int) \ini_get('session.cache_expire')));
        }

        return true;
    }
```

Because of my config the Header is set to "no-store, no-cache, must-revalidate". That's almost fine for me. But the second Header is send by the Response. When i only use "new Symfony\Component\HttpFoundation\Response()" the cache-control-header "max-age=0, must-revalidate, private" is send. I have to add the following two lines to remove the second header:

```php
$response->headers->remove('Cache-Control');
$response->headers->set(AbstractSessionListener::NO_AUTO_CACHE_CONTROL_HEADER, '');
```

Is this really correct, or is there another solution to just generate one header?

**Comment by MatTheCat** [comment]
Could you show your `framework.session` configuration? By default `AbstractSessionHandler`’s header should not be sent because `cache_limiter` is set to `0`.

(Also, make sure you didn’t override the `session.storage.options` parameter.)

**Reply by magicW** [reply]
my `cache_limiter `option is set to `nocache`. This is to get the header `no-store, no-cache, must-revalidate`. This is the header i want to have.

I just don't understand why the default Response send another header. I can't imagine this being a rare scenario. So I think I'm doing something wrong.

**Reply by MatTheCat** [reply]
Well you're not really supposed to change `cache_limiter`ʼs value (thatʼs why itʼs a parameter rather than configuration) . By default the response will send the header you want!

**Reply by magicW** [reply]
When I delete my `cache_limiter` config, it is `""` and not `0`. As far as I can see, this comes from the constructor of the NativeSessionStorage:

```php
public function __construct(array $options = [], $handler = null, ?MetadataBag $metaBag = null)
    {
        if (!\extension_loaded('session')) {
            throw new \LogicException('PHP extension "session" is required.');
        }

        $options += [
            'cache_limiter' => '',
            'cache_expire' => 0,
            'use_cookies' => 1,
            'lazy_write' => 1,
            'use_strict_mode' => 1,
        ];

        session_register_shutdown();

        $this->setMetadataBag($metaBag);
        $this->setOptions($options);
        $this->setSaveHandler($handler);
    }
``` 

And with this config, the AbstractSessionHandler will send the header `Cache-Control: max-age=%d, private, must-revalidate` with the `open`-method-call. But i want to send a no-store-header.

Maybe I should set the cache_limiter-option to "0" and set the wanted header in the response object?

(If it is important: my symfony/http-foundation - version is 5.4.35)

**Reply by MatTheCat** [reply, marked as answer]
Okay, I guess you’re not using the FrameworkBundle? Else you’d need to override the `session.storage.options` parameter to not have `cache_limiter` equals to `'0'`, and you haven’t mentioned this.

If you’re instantiating `NativeSessionStorage` yourself you must do so with the `cache_limiter` option set to `'0'`, that way no header will be sent by an `AbstractSessionHandler`.

**Reply by magicW** [reply]
Thanks for your help with the analysis. You're right that i am instantiating the `NativeSessionStorage` myself (using Symfony's NativeSessionStorageFactory). I'm now using your recommendation and set the `cache_limiter `- Option to `"0"`.

It might be helpful if Symfony's `NativeSessionStorageFactory `sets this option itself if it is not mentioned in the options.

**Reply by MatTheCat** [reply]
Yeah I’ve got no idea why Symfony changed the empty string semantic; also see https://github.com/symfony/symfony/commit/cf4eb4602ace6c32523133b09feab79a53f8fa2d.
